ledger/participant-integration-api/src/main/scala/platform/store/backend/common/CommonStorageBackend.scala

```diff
@@ -809,43 +825,106 @@ private[backend] trait CommonStorageBackend[DB_BATCH] extends StorageBackend[DB_
               WHERE
                 archive_events.event_offset <= $pruneUpToInclusive AND
                 archive_events.contract_id = delete_events.contract_id
-            )""",
-      SQL"""
-          -- Create events (only for contracts archived before the specified offset)
-          delete from participant_events_create delete_events
+            )"""
+
+    val retroactiveDivulgencesPruning =
+      if (pruneAllDivulgedContracts)
+        SQL"""
+          -- Retroactive divulgence events
+          delete from participant_events_divulgence delete_events
+          where delete_events.event_offset <= $pruneUpToInclusive"""
+      else
+        SQL"""
+          -- Divulgence events (only for contracts archived before the specified offset)
+          delete from participant_events_divulgence delete_events
           where
             delete_events.event_offset <= $pruneUpToInclusive and
             exists (
               SELECT 1 FROM participant_events_consuming_exercise archive_events
               WHERE
                 archive_events.event_offset <= $pruneUpToInclusive AND
                 archive_events.contract_id = delete_events.contract_id
-            )""",
-      SQL"""
-          -- Exercise events (consuming)
-          delete from participant_events_consuming_exercise delete_events
-          where
-            delete_events.event_offset <= $pruneUpToInclusive""",
-      SQL"""
-          -- Exercise events (non-consuming)
-          delete from participant_events_non_consuming_exercise delete_events
-          where
-            delete_events.event_offset <= $pruneUpToInclusive""",
-    ).foreach(_.execute()(connection))
+            )"""
+
+    {
+      conditionalImmediateDivulgencePruning ++
+        List(
+          retroactiveDivulgencesPruning,
+          createsPruning,
+          consumingExercisesPruning,
+          nonConsumingExercisesPruning,
+        )
+    }.foreach(_.execute()(connection))
+  }
+
+  // TODO divulgence pruning: Check if specialization for Oracle is needed (as it is for H2)
+  protected[CommonStorageBackend] def pruneImmediateDivulgence(
+      pruneUpToInclusive: Offset,
+      pruneAllDivulgedContracts: Boolean,
+  ): List[SimpleSql[Row]] = {
+    import com.daml.platform.store.Conversions.OffsetToStatement
+
+    if (pruneAllDivulgedContracts)
+      List(
+        SQL"""
+          -- Immediate divulgence events
+          WITH
+          	creates_stakeholders AS (
+          		SELECT event_offset, event_sequential_id, unnest(flat_event_witnesses) AS stakeholder
+          		FROM participant_events_create
+          		WHERE event_offset <= $pruneUpToInclusive
+          	),
+            to_be_pruned AS(
+              SELECT DISTINCT(event_sequential_id) AS event_sequential_id
+              FROM creates_stakeholders
+              WHERE NOT EXISTS (
+              	SELECT 1
+              	FROM
+              		creates_stakeholders c INNER JOIN parties p ON c.stakeholder=p.party
```

Comment:
Why the join here? You already have the creates_stakeholders available outside?

Reply:
Also: consider moving the unnest of the flat_event_witnesses into this join here. This way you don't need the creates_stakeholders at all. you can directly iterate over the create events. Which then actually also allows you to inline the condition into the DELETE

Reply:
Updated according to the discussion had offline.